[#148] Feat: 짤 업로드 페이지의 TagSearchForm 구현

HY219 opened the pull request:
## 📝 작업 내용

- 새 태그 생성
  - 태그 검색 창에 새로운 태그 작성 후, enter || 검색 버튼 클릭 시, usePostTag api 훅으로 새 태그 생성
  - 자동완성 목록에 일치하는 tagName이 없을 경우, 새 태그 생성 api 요청
  - 검색 창에서 태그를 작성 후, enter || 검색 버튼 클릭 시
    - 새 태그의 tagId 경우, 새 태그 생성 api의 response로 tagId를 받아 전달
    - 기존 태그의 tagId 경우, 자동완성 목록의 첫번째 요소의 tagId를 전달
      (기존 태그의 경우, 자동완성 목록의 첫번째 요소와 동일할 것이다.)
  - 자동완성 목록에서 클릭했을 시
    - 기존 태그의 tagId 경우, 클릭한 요소의 tagId를 전달

- 짤 업로드 시 선택된 태그의 tagId 전달
  - 태그 추가 가능 개수를 10개 이하로 지정
    - `MAX_SEARCH_TAG_UPLOAD = 10`

- 전역 상태의 를 이용해 짤 업로드 페이지 추천 태그 출력
  - 기존의 RecommandTag 컴포넌트 import 제거
  - 짤 업로드 페이지의 추천 태그 = popularTags
  - 추천 태그가 페이지에 바로 보이는 것이 아닌, 검색창 클릭 시 목록으로 볼 수 있도록 수정됨

## 📷 스크린샷 (선택)

## 💬 리뷰 요구사항(선택)

- 기능 완료를 위해 기능이 정상 작동한다면, 머지 후 리팩토링을 진행해도 좋을 것 같습니다:)
  - `tagId`와 `tagName`의 상태관리를 따로 두고 있어, 함께 관리하도록 리팩토링 할 수 있을 것 같습니다.
  - `UploadTagBadge` 컴포넌트를 지금처럼 따로 생성하지 않고, 기존의 `TagBadge` 컴포넌트를 재사용하도록 리팩토링 할 수 있을 것 같습니다.
  => 위 두 가지에 대해 리팩토링이 어려울 것 같다면 말씀 부탁으려요:)
  
- 재민님과 익님께서 기존에 구현하신 `TagSearchForm, TagAutoCompelete, TagBadge` 컴포넌트와 코드가 동일한 부분이 있어, commits에서 추가된 코드를 확인하고 파악하는데 더 편리할 것 같습니다:)

## 📍 기타 (선택)

- TagSearchForm과 TagAutoComplete 파일에서 동일한 동작을 하는 코드를 재사용하기 위해 코드를 동일하게 가져왔습니다.
  - UploadTagSearchForm과 UploadTagAutoComplete 파일 생성

 -  TODO: [2024-03-17] Strict 모드는 개발 모드에서만 활성화되므로, 추후 삭제해야합니다. (React.StrictMode로 인해, 개발자 모드에서 useEffect가 두번 실행되므로 작성해주었습니다.)
    ```
    if (!isChangeState.current) {
      isChangeState.current = true;
      return;
    }
    ```



close #148

yjc2021 committed:
Style: TagSearchForm, TagAutoComplete style 수정

yjc2021 committed:
Style: 채팅 레이아웃 v3 적용

yjc2021 committed:
Cleanup: my-liked-zzals Outlet에서 TagSearchForm 제거

yjc2021 committed:
Style: chat layout v3 적용

yjc2021 committed:
Style: TagSearchForm v3 적용

yjc2021 committed:
Style: 헤더 및 드롭다운 v3 적용

yjc2021 committed:
Cleanup: MyUploadedZzals에서 TagSearchForm 제거

yjc2021 committed:
Style: 좋아요한 짤 페이지 masonry w-full로 수정

yjc2021 committed:
Feat: 모바일 navigation 구현

yjc2021 committed:
Modify: TagSearchForm 기본 크기 수정

yjc2021 committed:
Modify: layout without chat 설정 및 업로드 페이지 layout 적용

yjc2021 committed:
Refactor: Navigation Footer 컴포넌트화

yjc2021 committed:
Style: Chat toggle 버튼 스타일 수정

yjc2021 committed:
Style: 업로드 컴포넌트 반응형 스타일 추가

thalassophilia committed:
Modify: ref 제거, blur를 사용하여 자동완성 제어

thalassophilia committed:
Feat: 추천 태그를 아톰으로 관리

yjc2021 committed:
Modify: layout-with-chat 레이아웃 정책 적용

yjc2021 committed:
Rename: layout-without-chat 제거 후 routes 구조 원상복구

thalassophilia committed:
Modify: 클릭 가능한 태그는 x 버튼 추가

thalassophilia committed:
Chore: bg black 추가

thalassophilia committed:
Modify: 추천 태그를 아톰으로 변경

thalassophilia committed:
Modify: 추천 태그와 태그 선택 로직 수정

thalassophilia committed:
Modify: use suspense query로 수정

thalassophilia committed:
Merge branch '#112/style/layout-v3', remote-tracking branch 'origin' into #121/modify/tag-autocomplete

thalassophilia committed:
Modify: tagSearchForm 컴포넌트 위치 변경

thalassophilia committed:
Feat: badge key 추가

thalassophilia committed:
Style: badge 텍스트 크기 미치 패딩 수정

thalassophilia committed:
Style: 선택된 태그 스타일 수정

thalassophilia committed:
Feat: 키보드 이벤트 추가

thalassophilia committed:
Feat: escape 키 누를시 자동완성 영역 close

thalassophilia committed:
Modify: 초기화 버튼 수정

thalassophilia committed:
Modify: 자동완성 키보드 영역 로직 수정

thalassophilia committed:
Modify: input 값이 태그에 포함되면 추가

thalassophilia committed:
Modify: tag 클릭 이벤트 제거

thalassophilia committed:
Merge remote-tracking branch 'origin' into #121/modify/tag-autocomplete

thalassophilia committed:
Modify: 자동완성에 없는 태그여도 검색 가능하게 수정

thalassophilia committed:
Style: tag badge 패딩 수정

thalassophilia committed:
Style: tag badge 글자 크기 및 배치 수정

HY219 committed:
Rename: 코드 재사용을 위해 UploadTagSearchForm (기존 TagSearchForm 코드 재사용)

./src/components/UploadZzal 에 파일 추가
UploadTagSearchForm과 UploadTagAutoComplete 파일 생성
TagSearchForm과 TagAutoComplete 파일에서 동일한 동작을 하는 코드를 재사용하기 위해 코드를 동일하게 가져왔습니다.

HY219 committed:
Feat: 전역 상태의  를 이용해 짤 업로드 페이지 추천 태그 출력

기존의 RecommandTag 컴포넌트 import 제거
짤 업로드 페이지의 추천 태그 = popularTags
추천 태그가 페이지에 바로 보이는 것이 아닌, 검색창 클릭 시 목록으로 볼 수 있도록 수정됨

HY219 committed:
Feat: 태그 추가 가능 개수를 10개 이하로 지정

MAX_SEARCH_TAG_UPLOAD = 10

HY219 committed:
Feat: 새 태그 생성 및 짤 업로드 시 선택된 태그의 tagId 전송

태그 검색 창에 새로운 태그 작성 후, enter || 검색 버튼 클릭 시, usePostTag api 훅으로 새 태그 생성
자동완성 목록에 일치하는 tagName이 없을 경우, 새 태그 생성 api 요청
새 태그의 tagId 경우, 새 태그 생성 api의 response로 tagId를 받아 전달
기존 태그의 tagId 경우, 검색 창에서 작성했을 시, 자동완성 목록의 첫번째 요소의 tagId를 전달
(기존 태그의 경우, 자동완성 목록의 첫번째 요소와 동일할 것이다.)
기존 태그의 tagId 경우, 자동완성 목록에서 클릭했을 시, 클릭한 요소의 tagId를 전달

hyeonjinan096 reviewed (commented):
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/107539614/ae871e44-f430-47c6-b4c9-2fc084a334a8)
ui는 아직 안된 걸까요 추후 수정해주시면 좋을 것 같습니다:)

hyeonjinan096 commented on src/components/UploadZzal/UploadTagSearchForm.tsx:
```diff
@@ -0,0 +1,188 @@
+            onChange={handleChangeTagInput}
+            autoComplete="off"
+            className="z-20 min-h-12 flex-1 rounded-xl border-none bg-transparent py-6pxr outline-none"
+          />
```
글자 수 제한이 있으면 좋을 것 같습니당

hyeonjinan096 commented on src/components/UploadZzal/UploadTagBadge.tsx:
```diff
@@ -0,0 +1,44 @@
+  return (
+    <span
+      className={cn(
+        "flex-column flex w-fit items-center rounded-3xl bg-black p-1 text-[14px] font-bold text-white",
```
flex-col로 변경하면 좋을 것 같습니당

HY219 replied:
`flex-col`을 두면 UI가 깨져서 제거해줘야할 것 같네요 :)
![image](https://github.com/zzalmyu/zzalmyu-frontend/assets/82134668/12934435-3008-435d-9f0d-63bd41b87228)

HY219 commented:
> ![image](https://private-user-images.githubusercontent.com/107539614/313476716-ae871e44-f430-47c6-b4c9-2fc084a334a8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2ODUzOTYsIm5iZiI6MTcxMDY4NTA5NiwicGF0aCI6Ii8xMDc1Mzk2MTQvMzEzNDc2NzE2LWFlODcxZTQ0LWY0MzAtNDdjNi1iNGM5LTJmYzA4NGEzMzRhOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMxN1QxNDE4MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YTIyZDhiZmUwMjE5Yjk2MWM3OGE4MmY2M2JiMDZmZDE0YjE5MWI5MGZlYmNiMTc2MDg3Zjg5YjY1YjA3YzM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yNPeQ3Lur-MEUV6BByjX8yOuGQqYOC3RvtKnNWXG-Pw) ui는 아직 안된 걸까요 추후 수정해주시면 좋을 것 같습니다:)

이 부분은 지금 수정해야 할 것 같네요. 수정하도록 하겠습니다!

HY219 committed:
Style: 선택한 태그와 업로드하기 버튼이 겹치지 않도록 UI 수정

임의로 magin값으로 적용했습니다.

choi-ik commented on src/components/UploadZzal/UploadTagAutoComplete.tsx:
```diff
@@ -0,0 +1,102 @@
+      !selectedTags.find((tag) => tag.tagId === tagId && tag.tagName === tagName)
+    ) {
+      setSelectedTags((previousState) => [...previousState, { tagId, tagName }]);
+      console.log(selectedTags);
```
console 지워주세욧

HY219 replied:
반영했습니다:)

choi-ik commented on src/components/UploadZzal/UploadTagAutoComplete.tsx:
```diff
@@ -0,0 +1,102 @@
+          </div>
+          <ul className="flex-column mb-10pxr flex flex-wrap gap-6pxr">
+            {selectedTags.map(({ tagId, tagName }, index) => (
+              <li key={`${index}-${tagId}`}>
```
key는 tagID만 주어도 될 것 같습ㄴ디ㅏ!

HY219 replied:
변경했습니다:)

choi-ik commented on src/components/UploadZzal/UploadTagSearchForm.tsx:
```diff
@@ -0,0 +1,188 @@
+  const { createTag } = usePostTag();
+  const [recommendedTags] = useAtom($recommendedTags);
+  const [selectedTags, setSelectedTags] = useAtom($selectedTagsUpload);
+  const isMounted = useRef(false);
```
추후에 네이밍을 Ref를 부여한 엘리먼트 네이밍으로 변경해주면 좋을 것 같습니다 !

HY219 replied:
우선은 isMountedRef로 변경해주었습니다!

HY219 committed:
Style: flex-column 제거

flex-col 적용 시 UI 깨짐
잘 못 작성된 flex-column 코드 제거

HY219 committed:
Refactor: 자동완성 목록 key를 tagId로만 지정 및 Ref 부여한 네이밍으로 변경

tagId로도 고유하므로 index 제거
console.log 제거
isMounted -> isMountedRef
isChangeState -> isChangeStateRef

HY219 committed:
Merge branch 'main' into #148/feature/tag-search-and-create

HY219 committed:
Merge branch 'main' of https://github.com/zzalmyu/zzalmyu-frontend into

HY219 committed:
Merge branch '#148/feature/tag-search-and-create' of https://github.com/zzalmyu/zzalmyu-frontend into #148/feature/tag-search-and-create

HY219 committed:
Merge: 기존 생성한 usePostUsedTag 커스텀 api 훅 추가

thalassophilia commented:
<img width="1206" alt="스크린샷 2024-03-18 오전 1 19 32" src="https://github.com/zzalmyu/zzalmyu-frontend/assets/61978339/c87791eb-e5b0-4f36-9d7b-b6eef806057e">
에러 로그입니다!

HY219 committed:
Fix: main 머지 시 제거되었던 postTag 코드 생성

태그 생성 요청 api

yjc2021 commented:
지금 JPG 파일만 업로드가 가능한데 png, gif까지 지원할 수 있는지 조사해봐야 할 것 같습니다

HY219 commented:
기능 우선 머지로 머지하겠습니다:)

추후 리팩토링이 필요할 수 있습니다.
> 지금 JPG 파일만 업로드가 가능한데 png, gif까지 지원할 수 있는지 조사해봐야 할 것 같습니다

> 추후에 네이밍을 Ref를 부여한 엘리먼트 네이밍으로 변경해주면 좋을 것 같습니다 !
-> 우선은 isMountedRef로 변경해주었습니다!

> 태그 글자 수 제한이 있으면 좋을 것 같습니당

> 전체 UI 수정 및 "업로드하기"버튼 css 속성 수정
